components/script/dom/webglrenderingcontext.rs

```diff
@@ -364,6 +369,77 @@ impl WebGLRenderingContext {
+    //      arguments, and then transferred to the WebGL
+    //      implementation. If a packed pixel format is specified
+    //      which would imply loss of bits of precision from the image
+    //      data, this loss of precision must occur."
+    fn rgba8_image_to_tex_image_data(&self,
+                                     format: TexFormat,
+                                     data_type: TexDataType,
+                                     pixels: Vec<u8> ) -> Vec<u8> {
+        // hint for vector allocation sizing.
+        let pixel_count = pixels.len() / 4;
+
+        match (format, data_type) {
+            (TexFormat::RGBA, TexDataType::UnsignedByte) => pixels,
+
+            (TexFormat::RGBA, TexDataType::UnsignedShort4444) => {
+                let mut rgba4 = Vec::<u8>::with_capacity(pixel_count * 2);
+                for rgba8 in pixels.chunks(4) {
+                    // FINISHME: These should be swapped on a big-endian system.
+                    rgba4.push((rgba8[3] & 0xf0) >> 4 |
+                               (rgba8[2] & 0xf0));
+                    rgba4.push((rgba8[1] & 0xf0) >> 4 |
+                               (rgba8[0] & 0xf0));
+                }
+                rgba4
+            }
+
+            (TexFormat::RGBA, TexDataType::UnsignedShort5551) => {
+                let mut rgba5551 = Vec::<u8>::with_capacity(pixel_count * 2);
+                for rgba8 in pixels.chunks(4) {
+                    let pix =
+                        rgba8[3] as u16 >> 7 |
+                        (rgba8[2] as u16 & 0xf8) >> 2 |
+                        (rgba8[1] as u16 & 0xf8) << 3 |
+                        (rgba8[0] as u16 & 0xf8) << 8;
+                    // FINISHME: These should be swapped on a big-endian system.
+                    rgba5551.push((pix & 0xff) as u8);
+                    rgba5551.push((pix >> 8) as u8);
+                }
+                rgba5551
+            }
+
+            (TexFormat::RGB, TexDataType::UnsignedShort565) => {
+                let mut rgb565 = Vec::<u8>::with_capacity(pixel_count * 2);
+                for rgba8 in pixels.chunks(4) {
+                    let pix =
+                        (rgba8[2] as u16 & 0xf8) >> 3 |
+                        (rgba8[1] as u16 & 0xfc) << 3 |
+                        (rgba8[0] as u16 & 0xf8) << 8;
+                    // FINISHME: These should be swapped on a big-endian system.
+                    rgb565.push((pix & 0xff) as u8);
+                    rgb565.push((pix >> 8) as u8);
+                }
+                rgb565
+            }
+
+            // Validation should have ensured that we only hit the
+            // above cases, but we haven't turned the (format, type)
+            // into an enum yet so there's a default case here.  All
+            // WebGL types are 4 bytes per pixel or smaller, so just
+            // return the incoming RGBA8 pixels in this path.
+            _ => pixels
```

Comment:
OK, looks like unreachable!() should always panic.  I thought I'd seen something in my other debugging that suggested that it didn't panic.